data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginBeanFactoryProvider.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugin;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.support.GenericApplicationContext;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import java.util.Objects;
+
+/**
+ * @since 1.3
+ * <p>
+ *     Used to create new instances of ApplicationContext that can be used to provide a per plugin instance isolated ApplicationContext
+ *     scope. CoreApplicationContext is unavailable to sharedPluginApplicationContext and its children.
+ * </p>
+ * <p>pluginIsolatedApplicationContext inherits from {@link PluginBeanFactoryProvider#sharedPluginApplicationContext}</p>
+ * <p>{@link PluginBeanFactoryProvider#sharedPluginApplicationContext} inherits from <i>publicContext</i></p>
+ * <p><i>publicContext</i> is the root {@link ApplicationContext}</p>
+ */
+@Named
+class PluginBeanFactoryProvider implements Provider<BeanFactory> {
+    private static final Logger LOG = LoggerFactory.getLogger(PluginBeanFactoryProvider.class);
+    private final ApplicationContext sharedPluginApplicationContext;
+
+    @Inject
+    PluginBeanFactoryProvider(final ApplicationContext coreContext) {
+        final ApplicationContext publicContext = Objects.requireNonNull(coreContext.getParent());
+        sharedPluginApplicationContext = new GenericApplicationContext(publicContext);
+    }
+
+    /**
+     * @since 1.3
+     * Creates a new Application context that inherits from {@link PluginBeanFactoryProvider#sharedPluginApplicationContext} then returns
+     * new context's BeanFactory.
+     * @return BeanFactory A BeanFactory that inherits from {@link PluginBeanFactoryProvider#sharedPluginApplicationContext}
+     */
+    public BeanFactory get() {
+        final GenericApplicationContext pluginIsolatedApplicationContext = new GenericApplicationContext(sharedPluginApplicationContext);
```

Comment:
Why does the `pluginIsolatedApplicationContext` need to be created during `get()`? Can this be created as a class variable instead of `sharedPluginApplicationContext`?

Reply:
Thanks for the close attention to variable scope. I completely agree, especially when DI is involved, polluting a shared scope is frequently an issue.

Updating `get()` to return a class variable will reduce the level of isolation for plugins. If a class variable is used every plugin instance will have the same mutable context.

In the current proposal, `get()` creates a new context providing each plugin instance an isolated context. 

An example of how using a single instance class variable could be misused:
Given there are two OpenSearch sinks defined in a configuration. Using a shared ApplicationContext to provide auth credential Beans, all instances of the sink would be provided the same instance of a credential Bean. The credential Bean is also be exposed to all plugins.

Given the credential bean scope is changed to provide a new instance on each injection. Each OpenSearch sink would need to configure the credential bean from data held by the sink bean. The approach reduces the benefit gained from using DI in my opinion.

Let me know your thoughts and if you have ideas on how else we could implement the context hierarchy.